File: core/src/main/kotlin/br/com/dillmann/nginxignition/core/common/validation/ConsistencyValidator.kt
package br.com.dillmann.nginxignition.core.common.validation

typealias ErrorCreator = (String, String) -> Unit

internal abstract class ConsistencyValidator {
    protected suspend fun withValidationScope(scope: suspend (addError: ErrorCreator) -> Unit) {
        val errors = mutableListOf<ConsistencyException.Violation>()
        val addError: ErrorCreator = { path, message ->
            errors += ConsistencyException.Violation(path, message)
        }

        scope(addError)

        if (errors.isNotEmpty()) {
            throw ConsistencyException(errors)
        }
    }
}

File: core/common/validation/consistency_validator.go
package validation

import (
	"dillmann.com.br/nginx-ignition/core/common/i18n"
)

type ConsistencyValidator struct {
	violations []ConsistencyViolation
}

func NewValidator() *ConsistencyValidator {
	return &ConsistencyValidator{}
}

func (v *ConsistencyValidator) Add(path string, message *i18n.Message) {
	v.violations = append(v.violations, ConsistencyViolation{
		Path:    path,
		Message: message,
	})
}

func (v *ConsistencyValidator) Result() error {
	if len(v.violations) == 0 {
		return nil
	}

	return NewError(v.violations)
}
